capstone/capapi/middleware.py

```diff
@@ -92,3 +93,99 @@ def process_request(self, request):
+            return response
+
+        # parse the range(s) -- at the moment, we're only responding to the
+        # first range, but we coalesce adjacent and overlapping ranges below
+        http_range = request.META.get('HTTP_RANGE')
+        units = "bytes="
+        if http_range and http_range.startswith(units):
+            ranges = []
+            matches = re.finditer(r"(\d*)-(\d*),?", http_range[len(units):])
+            if matches:
+                # treat the response as a file so we can see how big it is
+                f = response.file_to_stream
+                # response_size = os.fstat(f.fileno()).st_size
+                response_size = f.size
+            for match in matches:
+                start = match.group(1)
+                end = match.group(2)
+                # the case where both ends of the range are empty
+                if not start and not end:
+                    continue
+                # the case where start is empty
+                if not start:
+                    start = max(0, response_size - int(end))
+                    end = response_size - 1
+                else:
+                    start = int(start)
+                # this handles the case where end is empty and where it isn't
+                end = int(end or response_size - 1)
+                if start >= response_size or start > end:
+                    response.status_code = 416
+                    response['Content-Range'] = 'bytes */%d' % response_size
+                    return response
+                end = min(end, response_size - 1)
+                ranges.append((start, end))
+        else:
+            return response
+
+        # we're not really handling If-Range headers; if present and ill-formed,
+        # return the whole content
+        if_range = request.META.get('HTTP_IF_RANGE')
+        if if_range and if_range != response.get('Last-Modified') and if_range != response.get('ETag'):
+            return response
+
+        # coalesce adjacent and overlapping ranges
+        collapsed_ranges = []
+        for cur in sorted(ranges):
+            if len(collapsed_ranges) == 0:
+                collapsed_ranges.append(cur)
+            else:
+                prev = collapsed_ranges.pop()
+                if cur[0] - 1 <= prev[1]:
+                    collapsed_ranges.append((prev[0], max(prev[1], cur[1])))
+                else:
+                    collapsed_ranges.append(prev)
+                    collapsed_ranges.append(cur)
+        ranges = collapsed_ranges
+
+        # to handle more than the first range, we'd need to produce a
+        # multipart response...
+        if len(ranges) > 0:
+            (start, end) = ranges[0]
+
+            # if the range encompasses the whole response, return 200 --
+            if start == 0 and end >= (response_size - 1):
+                return response
+
+            # iterator for range of file
+            def fchunks(start, end):
```

Comment:
This should use `seek` and much larger chunk sizes for efficiency. Something along the lines of:

```
def fchunks(start, end):
    remaining = end - start
    f.seek(start)
    while remaining > 0:
        chunk = f.read(min(remaining, 2**20))
        if not chunk:
            break
        yield chunk
        remaining -= len(chunk)
```

Though that may have an off-by-one error -- not sure I used start and end correctly.